### PR TITLE
Removal of API endpoint & modification to gRPC auth

### PR DESCRIPTION
**Pull request by JosteinLindhom**
- Fetch self through gRPC request
- Updated solution to work with removal of API endpoint.
- Removed GetSelf & JSONuser struct.

### CODE REVIEW

File: web/auth/auth.go [hunk 1]
```diff
@@ -17,7 +17,7 @@ import (
 
 func init() {
 	gob.Register(&UserSession{})
-	TokenStore = &UserToken{store: make(map[string]uint64)}
+	TokenStore = &UserToken{store: make(map[string]*http.Request)}
```
Comment (meling) [inline web/auth/auth.go:20]: This can be initialized at the point of declaration of `TokenStore`.

File: web/auth/auth.go [hunk 2]
```diff
@@ -49,20 +49,27 @@ func newUserSession(id uint64) *UserSession {
...
+	if sess, err := gothic.Store.Get(ut.store[token], SessionKey); err == nil {
+		if i, ok := sess.Values[UserKey]; ok {
+			us := i.(*UserSession)
+			return us.ID
+		}
+	}
+	return 0
 }
 
 var TokenStore *UserToken
```
Comment (meling) [inline web/auth/auth.go:75]: Suggest to move this line between the `UserToken` struct declaration and the `add` method.

File: web/users.go [hunk 3]
```diff
@@ -1,45 +1,9 @@
...
-	ID        uint64 `json:"id"`
-	IsAdmin   *bool  `json:"isadmin"`
-	Name      string `json:"name"`
-	StudentID string `json:"studentid"`
-	Email     string `json:"email"`
-	AvatarURL string `json:"avatarurl"`
-}
-
-// GetSelf redirects to GetUser with the current user's id.
-func GetSelf(db database.Database) echo.HandlerFunc {
```
Comment (meling) [inline web/users.go:23]: There is a `TestGetSelf` in `web/users_test.go` that now fails to compile. Can you rewrite `TestGetSelf` to work with the gRPC approach? You would probably need to call the `UserVerifier` interceptor or something, putting a dummy cookie/token into the context metadata, and checking that you get logged in. And similarly, that if you don't provide the required cookie, you won't be logged in.

File: public/src/managers/ServerProvider.ts [hunk 4]
```diff
@@ -301,20 +301,13 @@ export class ServerProvider implements IUserProvider, ICourseProvider {
...
-        const iusr = result.data;
-        const usr = new User();
-        usr.setId(iusr.id);
-        usr.setStudentid(iusr.studentid);
-        usr.setName(iusr.name);
-        usr.setEmail(iusr.email);
-        usr.setAvatarurl(iusr.avatarurl);
-        usr.setIsadmin(iusr.isadmin);
-        return usr;
+        const iusr = result.data
```
Comment (meling) [inline public/src/managers/ServerProvider.ts:309]: Here you can just `return result.data`

**Comment by meling** [comment]
Is this ready to be reviewed in more detail? It is easy to fix the conflicting files mentioned above. You can just do:
```shell
git checkout master
git pull
gh pr checkout 461
git merge master
<fix the conflicting files>
git add <updated files>
git commit
git push
```

**Comment by meling** [comment]
@JosteinLindhom ping :-)

File: web/auth/auth.go [hunk 5]
```diff
@@ -47,28 +51,25 @@ func newUserSession(id uint64) *UserSession {
...
+func (us *UserSession) enableProvider(provider string) {
+	us.Providers[provider] = struct{}{}
 }
 
-func (ut *UserToken) add(token string, id uint64) {
-	for token, userID := range TokenStore.store {
-		if userID == id {
-			delete(TokenStore.store, token)
+// map from session cookies to user IDs.
+var tokenStore = make(map[string]uint64)
```
Comment (meling) [inline web/auth/auth.go:59]: Should we use the term `cookie` instead of `token` for this stuff? (including the `Add(token ...` method below)

**Review by meling — CHANGES_REQUESTED** [review]

File: web/auth/auth.go (hunk 5 above)
Comment (JosteinLindhom) [reply web/auth/auth.go:59]: Agreed, that would make more sense.

File: web/auth/auth.go [hunk 6]
```diff
@@ -310,8 +327,8 @@ func AccessControl(logger *zap.Logger, db database.Database, scms *Scms) echo.Mi
 
 			i, ok := sess.Values[UserKey]
 			if !ok {
-				logger.Error(echo.ErrUnauthorized.Error())
-				return echo.ErrUnauthorized
+				// logger.Error(echo.ErrUnauthorized.Error())
```
[the remark this answers is not in the document]
Comment (JosteinLindhom) [reply web/auth/auth.go:326]: It should be removed.

Side note: I activated AccessControl for the whole site. Users not logged in, and logged in users should be allowed access to the website. Access control for data retreival is handled by TokenStore and the gRPC interceptor.

**Pull request merged**